### PR TITLE
schedule() considers instruction map

### PR DESCRIPTION
**Pull request by MozammilQ**
### Summary
In issue #10837 , I discovered that the schedule of qiskit.compiler works fine for BackendV1 but doesn't work for BackendV2. See this [comment](https://github.com/Qiskit/qiskit/issues/10837#issuecomment-1750955255) for details.

### Details and comments
**Details**:
On inspection I discovered that measure function in macros.py requires backend, It uses backend to decide which of the two private methods _measure_v1 or _measure_v2 to call.
So, I decided to pass in this backend from the user to measure of macros.py
I did this as follows:
**1>>** I changed the definition of schedule_circuit in qiskit/qiskit/scheduler/schedule_circuit.py and added backend as Optional here.
**2>>** I used this backend and passed it to as_soon_as_possible and as_late_as_possible of qiskit/qiskit/scheduler/methods/basic.py, I also changed the definition of these two methods as_soon_as_possible and as_late_as_possible.
**3>>** I passed this backend to lower_gates by changing the definition of lower_gates in qiskit/qiskit/scheduler/lowering.py 
**4>>** I also changed the definition of the nested function get_measure_schedule of lowering.py.
**5>>**  Finally I passed in the backend to measure of macros.py

I have made backend in the definition of these changed methods Optional because I do not want the tests to fail, complaining the methods require backend as an additional argument.
I have reformatted the files that I changed and made sure that all the tests have passed! :) 

**Comments**:
qiskit/qiskit/compile/scheduler.py in the definition of schedule one cannot omit the backend because, how the inst_map is processed depends on it.Furthermore, there is no need to ask for inst_map and meas_map separately if you already have the backend with you!So, backend should not be Optional here! I think inst_map, meas_map and dt are redundant here, please correct me if I am wrong :)

Final result after issue fixed:
fix_issue_10837

### CODE REVIEW

**Comment by MozammilQ** [comment]
sir, please see if my efforts a good enough :), please let me know if any additional changes are required :)

**Comment by nkanazawa1989** [comment]
@to24toro Could you please review the PR first? I'll merge if this looks good.

File: qiskit/scheduler/lowering.py
Comment (to24toro) [inline qiskit/scheduler/lowering.py]: `get_measure_schedule` is a inner function and you can pass and use `backend`  without adding `backend`  as an argument.
```suggestion
    def get_measure_schedule(qubit_mem_slots: Dict[int, int]) -> CircuitPulseDef:
```

**Comment by to24toro** [comment]
Thank you @MozammilQ .
It almost is good for me and will help solving the issue of scheduling with backendV2.

I have two additional requests for improvement.
- Add an argument of `backend` for the `sequence` in https://github.com/Qiskit/qiskit/blob/947e175edca77900cf1b0b266c55361bca0a7add/qiskit/scheduler/sequence.py#L28 and `lower_gates` in this `sequence`.
- Include explanations for added arguments to pass tests lint.

**Review by to24toro — CHANGES_REQUESTED** [review]

**Comment by MozammilQ** [comment]
> Thank you @to24toro for your suggestions.
> 
> I have done some improvements.
> 
>     * Add an argument of `backend` for the `sequence` in https://github.com/Qiskit/qiskit/blob/947e175edca77900cf1b0b266c55361bca0a7add/qiskit/scheduler/sequence.py#L28
**DONE**
>        and `lower_gates` in this `sequence`. 
**DONE**
> 
>     * Include explanations for added arguments to pass tests lint.
_I do not understand what I am supposed to do here, anyways I have added relevant docstring in the changed methods which I forgot to add, am I supposed to do anything else? adding explanations for added arguments in which file? and I have run all tests with tox and all tests have passed._

**Comment by to24toro** [comment]
Thank you @MozammilQ . I will confirm it.
Anyway, you need to pass all the test to merge it (see :https://dev.azure.com/qiskit-ci/qiskit/_build/results?buildId=50336&view=logs&jobId=86d13450-b8b4-5d9a-22f1-9f7e71062a71&j=86d13450-b8b4-5d9a-22f1-9f7e71062a71&t=a7f62dd9-9d78-5db3-3068-aed07887bd76). Could you please run `tox -eblack` and `tox -elint` in your PC and add the modification, referring to it?

**Comment by MozammilQ** [comment]
@to24toro , :)
Aaah!! Now, I got it! thanks for the kind words, you explained it very well.
I ran `tox -eblack` and `tox -elint`, and did the required modifications in the last commit. Is it good this time? :) 
just for the information I have run tox -eblack/tox -elint on fedora 38.

**Comment by to24toro** [comment]
Thank you @MozammilQ .
I think we are almost done. The left thing is to write test codes for backendV2 because the existing test codes related with your modification use backendV1. 

Especially, it's better to add the backendV2 version of https://github.com/Qiskit/qiskit/blob/65e8c1a7740be6bbd25c332da14df470f76d24d9/test/python/scheduler/test_basic_scheduler.py#L37 . These tests don't include handling the case that backend is backendV2.
However, this requires some knowledges and it might be tough.

File: qiskit/scheduler/methods/basic.py
Comment (to24toro) [inline qiskit/scheduler/methods/basic.py]:
```suggestion
        backend: A backend used to build the Schedule, the backend could be BackendV1
```

File: qiskit/scheduler/methods/basic.py
Comment (to24toro) [inline qiskit/scheduler/methods/basic.py]:
```suggestion
        backend: A backend used to build the Schedule, the backend could be BackendV1
```

File: qiskit/scheduler/schedule_circuit.py
Comment (to24toro) [inline qiskit/scheduler/schedule_circuit.py]:
```suggestion
        backend: A backend used to build the Schedule, the backend could be BackendV1
```

File: qiskit/scheduler/sequence.py
Comment (to24toro) [inline qiskit/scheduler/sequence.py]:
```suggestion
        backend: A backend used to build the Schedule, the backend could be BackendV1
```

**Review by to24toro — CHANGES_REQUESTED** [review]

**Comment by MozammilQ** [comment]
@to24toro , please wait :)

**Comment by TsafrirA** [comment]
@to24toro @MozammilQ 
No updates were made to any tests. Can't we add `FakeBackendV2` to the tests?

**Comment by MozammilQ** [comment]
@TsafrirA , I have added many tests that uses BackendV2, there are a few tests that I am yet to add in, I am in talks with @to24toro , regarding that :)

**Comment by MozammilQ** [comment]
> Thank you @MozammilQ . I think we are almost done. The left thing is to write test codes for backendV2 because the existing test codes related with your modification use backendV1.
> 
> Especially, it's better to add the backendV2 version of
> 
> https://github.com/Qiskit/qiskit/blob/65e8c1a7740be6bbd25c332da14df470f76d24d9/test/python/scheduler/test_basic_scheduler.py#L37
> . These tests don't include handling the case that backend is backendV2.
> However, this requires some knowledges and it might be tough.

@to24toro , please see if this is ok :)

File: test/python/scheduler/test_basic_scheduler.py
Comment (to24toro) [inline test/python/scheduler/test_basic_scheduler.py]:
```suggestion
```

This test is not related with the backend version and you can remove it.

File: test/python/scheduler/test_basic_scheduler.py
Comment (to24toro) [inline test/python/scheduler/test_basic_scheduler.py]: I don't understand that you change the number of gates from backendV1?
Isn't it enough to replace U2 gates with sx gates?

**Review by to24toro — CHANGES_REQUESTED** [review]
Thank you @MozammilQ for adding tests. Your comments about the pulse durations is very clear and helps me understand it. I suggested the minor changes.

I would like you to add the release note about the bugfix (see: https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#release-notes)

File: releasenotes/notes/Adding-BackendV2-support-to-schedule-of-qiskit.compiler.scheduler-a0366d3397c73af0.yaml
Comment (to24toro) [inline releasenotes/notes/Adding-BackendV2-support-to-schedule-of-qiskit.compiler.scheduler-a0366d3397c73af0.yaml]:
```suggestion
```
Thank you for adding the release notes and sorry for my poor explanation. It would be sufficient to just add an item for "fixes". Also, please describe what was fixed in the fixes section. It would be good to briefly mention what was fixed, such as enabling scheduling using backendV2.

**Review by to24toro — CHANGES_REQUESTED** [review]

**Comment by MozammilQ** [comment]
@to24toro , is everything good now? :)

**Review by to24toro — APPROVED** [review]
Thank you @MozammilQ . This Bugfix definitely helps users.

File: qiskit/scheduler/lowering.py
Comment (TsafrirA) [inline qiskit/scheduler/lowering.py:42]: It's a little nitpicky, but all these unions could be replaced with `Backend`.

File: releasenotes/notes/Adding-BackendV2-support-to-schedule-of-qiskit.compiler.scheduler-a0366d3397c73af0.yaml
Comment (nkanazawa1989) [inline releasenotes/notes/Adding-BackendV2-support-to-schedule-of-qiskit.compiler.scheduler-a0366d3397c73af0.yaml]:
```suggestion
features:
  - |
    Enables circuit to pulse scheduling using BackendV2.
    
    .. code-block:: python
    
      # import a fake backend which is a sub-class of BackendV2.
      from qiskit.providers.fake_provider import FakePerth
      from qiskit.compiler.scheduler import schedule
      from qiskit.circuit import QuantumCircuit
      
      qc = QuantumCircuit(1, 1)
      qc.x(0)
      qc.measure(0,0)
      sched = schedule(circuits=qc, backend=FakePerth())

    Since, BackendV2 was not supported by the :func:`~qiskit.compiler.schedule` function, this caused
    schedule method to error out when the :code:`backend` argument is supplied with :class:`.BackendV2`. 
    Refer to `#10837 <https://github.com/Qiskit/qiskit/issues/10837>` for more information.

```

Cleanup. I think this should be classified into new feature, because the PR enables new object input which has not been supported previously.

**Review by nkanazawa1989 — DISMISSED** [review]
Thanks @MozammilQ for contribution and @to24toro for review. I just added suggestion of the release note cleanup. You can push commit button if you like it. I'll approve the PR as is so that we can safely add this PR to merge queue before the release date.

**Comment by MozammilQ** [comment]
@nkanazawa1989 , does this look good now? :)

**Review by nkanazawa1989 — APPROVED** [review]
Thanks @MozammilQ this looks good to me. I'll add this to merge queue.

**Pull request merged**
